src/main/java/codeu/model/store/basic/ActivityStore.java

```diff
@@ -0,0 +1,146 @@
+package codeu.model.store.basic;
+
+import codeu.model.data.User;
+import codeu.model.data.Message;
+import codeu.model.data.Conversation;
+import codeu.model.data.Activity;
+import codeu.model.data.ActivityType;
+import codeu.model.store.persistence.PersistentStorageAgent;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.time.Instant;
+
+/**
+ * Store class that uses in-memory data structures to hold values and automatically loads from and
+ * saves to PersistentStorageAgent. It's a singleton so all servlet classes can access the same
+ * instance.
+ */
+public class ActivityStore {
+
+    /** Singleton instance of ActivityStore. */
+    private static ActivityStore instance;
+
+    /**
+     * Returns the singleton instance of ActivityStore that should be shared between all servlet classes.
+     * Do not call this function from a test; use getTestInstance() instead.
+     */
+    public static ActivityStore getInstance() {
+        if (instance == null) {
+            instance = new ActivityStore(PersistentStorageAgent.getInstance());
+        }
+        return instance;
+    }
+
+    /**
+     * Instance getter function used for testing. Supply a mock for PersistentStorageAgent.
+     *
+     * @param persistentStorageAgent a mock used for testing
+     */
+    public static ActivityStore getTestInstance(PersistentStorageAgent persistentStorageAgent) {
+        return new ActivityStore(persistentStorageAgent);
+    }
+
+    /**
+     * The PersistentStorageAgent responsible for loading Users from and saving Users to Datastore.
+     */
+    private PersistentStorageAgent persistentStorageAgent;
+
+    /** The in-memory list of Activities. */
+    private List<Activity> activities;
+
+    /** This class is a singleton, so its constructor is private. Call getInstance() instead. */
+    private ActivityStore(PersistentStorageAgent persistentStorageAgent) {
+        this.persistentStorageAgent = persistentStorageAgent;
+        activities = new ArrayList<>();
+    }
+
+     /**
+     * Access all Activity objects.
+     */
+    public List<Activity> getAllActivities() {
+        return activities;
+    }
+
+    /**
+     * Access all Activity objects, sorted from oldest to most recent.
+     * Sorts all of the activities based on creation attribute.
+     */
+    public List<Activity> getAllSortedActivities() {
+        List<Activity> activitiesCopy = new ArrayList<>(activities);
+        activitiesCopy.sort(Collections.reverseOrder());
+        return activitiesCopy;
+    }
+
+    /**
+     * Access all Activity objects that were created on or after a specified date.
+     * 
+     * @param instant time that the Activity objects must be created on or after
+     */
+    public List<Activity> getActivitiesMadeOnOrAfter(Instant instant) {
```

Comment:
Would that implementation be more efficient then the one we have currently? And it'll be sorted once I update getAllActivities to getAllSortedActivities (see above)...

Reply:
More efficient? Maybe. Its definitely not less efficient runtime wise. You are already creating a new list so it is not less space efficient (in fact it may be more efficient since you arent ever creating a longer list than you need). It is more readable though. For runtime it should be about the same, maybe more flexible because in the solution I am describing it doesn't need a sorted list because it would just go through the whole list once. Once you start with a sort its already less efficient than just doing a single pass of a list.

Granted if you do need/want it sorted the solution I am describing is pretty much the same as this one, just a bit more readable.

Reply:
Makes sense, I'll update it to the more readable form. I do want a sorted solution though, so I'll loop through the activities, populate the new list while doing so, then sort that, which should be the most efficient way?